actions/check_node_status.go

```diff
@@ -0,0 +1,120 @@
+package actions
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/cenkalti/backoff/v4"
+	"github.com/sirupsen/logrus"
+	"golang.org/x/sync/errgroup"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"k8s.io/client-go/kubernetes"
+
+	"github.com/castai/cluster-controller/castai"
+)
+
+func newCheckNodeStatusHandler(log logrus.FieldLogger, clientset kubernetes.Interface) ActionHandler {
+	return &checkNodeStatusHandler{
+		log:       log,
+		clientset: clientset,
+	}
+}
+
+type checkNodeStatusHandler struct {
+	log       logrus.FieldLogger
+	clientset kubernetes.Interface
+}
+
+func (h *checkNodeStatusHandler) Handle(ctx context.Context, data interface{}) error {
+	req, ok := data.(*castai.ActionCheckNodeStatus)
+	if !ok {
+		return fmt.Errorf("unexpected type %T for check node status handler", data)
+	}
+
+	log := h.log.WithFields(logrus.Fields{
+		"node_name":   req.NodeName,
+		"node_status": req.NodeStatus,
+	})
+
+	switch req.NodeStatus {
+	case castai.ActionCheckNodeStatus_READY:
+		log.Info("checking node ready")
+		return h.checkNodeReady(ctx, req)
+	case castai.ActionCheckNodeStatus_DELETED:
+		log.Info("checking node deleted")
+		return h.checkNodeDeleted(ctx, req)
+
+	}
+
+	return fmt.Errorf("unknown status to check provided node=%s status=%s", req.NodeName, req.NodeStatus)
+}
+
+func (h *checkNodeStatusHandler) checkNodeDeleted(ctx context.Context, req *castai.ActionCheckNodeStatus) error {
+	retries := uint64(5)
+	if req.WaitTimeoutSeconds != nil {
+		retries = uint64(*req.WaitTimeoutSeconds)
+	}
+	b := backoff.WithContext(backoff.WithMaxRetries(backoff.NewConstantBackOff(time.Second), retries), ctx)
+	return backoff.Retry(func() error {
+		n, err := h.clientset.CoreV1().Nodes().Get(ctx, req.NodeName, metav1.GetOptions{})
+		if apierrors.IsNotFound(err) {
+			return nil
+		}
+		if n != nil {
+			return backoff.Permanent(errors.New("node is not deleted"))
+		}
+		return err
+	}, b)
+}
+
+func (h *checkNodeStatusHandler) checkNodeReady(ctx context.Context, req *castai.ActionCheckNodeStatus) error {
+	timeout := 10 * time.Second
+	watchObject := metav1.SingleObject(metav1.ObjectMeta{Name: req.NodeName})
+	if req.WaitTimeoutSeconds != nil {
+		timeout = time.Duration(*req.WaitTimeoutSeconds) * time.Second
+	}
+	ctx, cancel := context.WithTimeout(context.Background(), timeout)
+	defer cancel()
+
+	timeoutInSec := int64(timeout.Seconds())
+	watchObject.TimeoutSeconds = &timeoutInSec
```

Comment:
why this is needed? we already have `context.WithTimeout(context.Background(), timeout)` so the watch should timeout anyway and setting multiple timeouts adds confussion